bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsXmppDevice.java

```diff
@@ -0,0 +1,463 @@
+                            parser.handleMessage(eventNameOpt.get(), iq.payload);
+                        } else {
+                            logger.debug("{}: Got unexpected XML payload {}", getSerialNumber(), iq.payload);
+                        }
+                    }
+                } catch (Exception e) {
+                    listener.onEventStreamFailure(EcovacsXmppDevice.this, e);
+                }
+            } else if (iqRequest instanceof ErrorIQ) {
+                StanzaError error = ((ErrorIQ) iqRequest).getError();
+                logger.trace("{}: Got error response {}", getSerialNumber(), error);
+                listener.onEventStreamFailure(EcovacsXmppDevice.this,
+                        new XMPPException.XMPPErrorException(iqRequest, error));
+            }
+            return null;
+        }
+    }
+
+    private static class CommandResponseHolder {
+        @Nullable
+        String response;
+    }
+
+    private static class DeviceCommandIQ extends IQ {
+        static final String TAG_NAME = "query";
+        static final String NAMESPACE = "com:ctl";
+
+        private final String payload;
+        final String id;
+
+        // request
+        public DeviceCommandIQ(IotDeviceCommand<?> cmd, Jid from, Jid to) throws Exception {
+            super(TAG_NAME, NAMESPACE);
+            setType(Type.set);
+            setTo(to);
+            setFrom(from);
+
+            this.id = createRequestId();
+            this.payload = cmd.getXmlPayload(id);
+        }
+
+        // response
+        public DeviceCommandIQ(@Nullable String id, String payload) {
+            super(TAG_NAME, NAMESPACE);
+            this.id = id != null ? id : "";
+            this.payload = payload.replaceAll("\n|\r", "");
+        }
+
+        @Override
+        protected @Nullable IQChildElementXmlStringBuilder getIQChildElementBuilder(
+                @Nullable IQChildElementXmlStringBuilder xml) {
+            if (xml != null) {
+                xml.rightAngleBracket();
+                xml.append(payload);
+            }
+            return xml;
+        }
+
+        private String createRequestId() {
+            // Ecovacs' app uses numbers for request IDs, so better constrain ourselves to that as well
+            int random8DigitNumber = 10000000 + new Random().nextInt(90000000);
+            return Integer.toString(random8DigitNumber);
+        }
+    }
+
+    private static class PingIQ extends SimpleIQ {
+        public PingIQ(Jid from, Jid to) {
+            super("ping", "urn:xmpp:ping");
+            setType(Type.get);
+            setFrom(from);
+            setTo(to);
+        }
+    }
+
+    private static class CommandIQProvider extends IQProvider<@Nullable DeviceCommandIQ> {
+        @Override
+        public @Nullable DeviceCommandIQ parse(@Nullable XmlPullParser parser, int initialDepth) throws Exception {
+            @Nullable
+            DeviceCommandIQ packet = null;
+
+            if (parser == null) {
```

Comment:
Null checking the parser should be unnecessary.

Reply:
I had to do that to keep the null checks happy. Since Smack's code does not have null annotations, it's assumed to be nullable, and the null checker then requires to check for null.

Reply:
Tried removing the `@Nullable` annotation for `parser` (which allows removing this check), but:

```
[ERROR] openhab-addons/bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsXmppDevice.java:[435,48] Illegal redefinition of parameter parser, inherited method from org.jivesoftware.smack.provider.Provider<org.openhab.binding.ecovacs.internal.api.impl.EcovacsXmppDevice.DeviceCommandIQ> does not constrain this parameter
```

... so I guess I'll have to keep it.